### ISSUE TITLE
Single-line fluent API for asserting an `event` has been raised

### ISSUE THREAD

**Issue by JayBazuzi** [fluentassertions/fluentassertions#423]
I like the `Invoking` / `Awaiting` syntax for `ShouldThrow()`, and wish for the same thing for events. Like this:

```
            var subject = new TestViewModel();
            subject.MonitoringEvents(
                _ => _.Value = "foo"
                ).ShouldRaisePropertyChangeFor(_ => _.Value);
```

Here's one way it could be implemented

```
    static class EventMonitoringAssertionExtensions
    {
        public static EventMonitoringAssertionHelper<T> MonitoringEvents<T>(this T subject, Action<T> action)
        {
            return new EventMonitoringAssertionHelper<T>(subject, action);
        }

        public class EventMonitoringAssertionHelper<T>
        {
            readonly Action<T> _action;
            readonly T _subject;

            public EventMonitoringAssertionHelper(T subject, Action<T> action)
            {
                _subject = subject;
                _action = action;
            }

            public void ShouldRaisePropertyChangeFor(Expression<Func<T, object>> propertyExpression)
            {
                _subject.MonitorEvents();
                _action(_subject);
                _subject.ShouldRaisePropertyChangeFor(propertyExpression);
            }
        }
    }
```

**Comment by dennisdoomen** [comment]
I like the idea, but somehow this doesn't read very fluently.

**Comment by somewhatabstract** [comment]
Perhaps we could do:

```
Action underTest = () -> DoSomething();

underTest.On(objectThatIsToBeMonitored).ShouldRaise( objectThatIsToBeMonitored => objectThatIsToBeMonitored.Event );
```

**Comment by dennisdoomen** [comment]
Or `subject.WhenMonitored().ShouldRaisePropertyChangedFor(o => o.property)`?

BTW, not getting why you need that `Action`.

**Comment by somewhatabstract** [comment]
Well, I think the point is that you are testing that a specific action leads to that event occurring. So you want to tell it what object to look at and what action to execute to raise the event.

However, I think if an Arrange/Act/Assert setup is used, then the act and the assert get separated out. I think the reason anyone asks for this is that unlike the rest of FluentAssertions, the monitoring has to be setup ahead of time so that it is monitoring when the act occurs.

By combining the act and the assertion, this becomes more fluent (but it also munges the act and the assert; so I'm torn as to whether it's valuable or not).

**Comment by JayBazuzi** [comment]
The "Arrange" step is about the system under test - configuring the system to be in a state where the "Action" makes sense.

But "MonitorEvents()" is not part of the the SUT - it's part of the assertion. It's weird/unfortunate that it must happen before the action.

Expressed as a strict AAA, the test should say:

```
    var foo = new Foo();

    foo.Bar();

    foo.ShouldHaveRaised(...);
```
But there's a technical constraint that prevents this from working. `Invoking().ShouldThrow()` has a similar constraint, and uses a similar pattern I've described.

**Comment by JayBazuzi** [comment]
Perhaps:

```
    foo.Invoking(f => f.Bar()).ShouldRaise...
```
